pkg/util/validation/errors.go

```diff
@@ -155,29 +183,40 @@ func (e *exemplarValidationError) Error() string {
 	return fmt.Sprintf(e.message, e.timestamp, mimirpb.FromLabelAdaptersToLabels(e.seriesLabels).String(), mimirpb.FromLabelAdaptersToLabels(e.exemplarLabels).String())
 }
 
-func newExemplarEmtpyLabelsError(seriesLabels []mimirpb.LabelAdapter, exemplarLabels []mimirpb.LabelAdapter, timestamp int64) ValidationError {
+var exemplarEmptyLabelsMsgFormat = globalerrors.Format(
+	globalerrors.ErrIDExemplarLabelsMissing,
+	"received exemplar with no valid labels, timestamp: %%d series: %%s labels: %%s")
```

Comment:
> series: %%s labels: %%s

It's probably better to wrap them by single quotes, for consistency.

Reply:
Instead of needing to double-escape placeholders here, could we return some sort of closure or object that formats when needed. Something like

```go
type ErrorFmt interface {
    Format(vals ... interface{}) string
}

var examplarEmptyLabelsMsgCallable = globalerrors.FormatCallable(
    globalerrors.ErrIDExemplarLabelsMissing,
    "received exemplar with no valid labels, timestamp: %d series: %s labels: %s")
    
someMessage := examplarEmptyLabelsMsgFormat.Format(ts, series, labels)
```

Not sure this is a huge improvement but it seems a little harder to use incorrectly.

Reply:
You're right: I haven't made it easy to use. I think I can remove the `%` escaping with a simple change to `Format()` (basically concatenating strings without using `fmt.Sprintf()`). I will give it a try to that first.

Reply:
Does this need to be plural? If not, please use singular:

```suggestion
	"received an exemplar with no valid label, timestamp: %%d series: %%s labels: %%s")
```

Reply:
> You're right: I haven't made it easy to use. I think I can remove the % escaping with a simple change to Format() (basically concatenating strings without using fmt.Sprintf()). I will give it a try to that first.

I looked at it again, and I personally think `Format()` is quite straightfoward. If you're building a template, it's quite obvious you have to escape `%` (when that needs to be part of the template).

I would like to hear other people opinion first.

Reply:
I would remove that extra formatting from `Format(format string, args ...interface{})` completely and rename it to `Message(msg string)`.

Most (well, I didn't count) of the usages here don't specify args, so they just wouldn't need to double escape (cool!), and in the places where we do actually provide a message, callers can `fmt.Sprintf()` themselves:
```go
func (e *invalidMetricNameError) Error() string {
	return globalerrors.Format(globalerrors.ErrIDInvalidMetricName, fmt.Sprintf("received series with invalid metric name: '%.200s'", e.metricName))
}
```

which would make it obvuous to them that they need to double escape if they want to format a formatting template:

```go
var exemplarMaxLabelLengthMsgFormat = globalerrors.Format(
	globalerrors.ErrIDExemplarLabelsTooLong,
	fmt.Sprintf("received exemplar whose combined labels set size exceeds the limit of %d characters, timestamp: %%d series: %%s labels: %%s", ExemplarMaxLabelSetLength),
)
```

Or, at least, I'd consider having both formatting and non-formatting versions, adding the formatting version `Formatf` (or `Messagef`, see proposal in a comment above) to the proper linter as a _formatting call_.

Reply:
Ok Oleg, you convinced me. Done.